server/src/main/java/io/crate/execution/dml/TranslogIndexer.java

```diff
@@ -85,19 +83,28 @@ public TranslogIndexer(DocTableInfo table, Version shardCreatedVersion) {
         this.indexColumns = buildIndexColumns(
             table.indexColumns(),
             table::getReference,
-            reference -> new IndexInput(reference.column(), () -> sourceMap)
+            reference -> new IndexInput(reference.column())
         );
         ignoreUnknownColumns = table.columnPolicy() != ColumnPolicy.STRICT;
         this.shardCreatedVersion = shardCreatedVersion;
     }
 
 
-    private record IndexInput(ColumnIdent columnIdent,
-                              Supplier<Map<String, Object>> sourceSupplier) implements Input<Object> {
+    private static class IndexInput implements Input<Object> {
+        private final List<String> fullPath;
+        private Map<String, Object> source;
+
+        IndexInput(ColumnIdent columnIdent) {
+            this.fullPath = Lists.concat(columnIdent.name(), columnIdent.path());
+        }
+
+        public void setSource(Map<String, Object> source) {
+            this.source = source;
+        }
 
         @Override
         public Object value() {
-            return Maps.getByPath(sourceSupplier.get(), Lists.concat(columnIdent.name(), columnIdent.path()));
+            return Maps.getByPath(source, fullPath);
```

Comment:
Moved path resolution to ctor to avoid calling `Lists.concat` per translog entry